### PR TITLE
VBLOCKS-1767 | Migrate AudioPlayer

### PR DESCRIPTION
**Pull request by charliesantos**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

## Pull Request Details

Please see https://issues.corp.twilio.com/browse/VBLOCKS-1767 for details

### Before review
* [x] Updated CHANGELOG.md if necessary
* [x] Added unit tests if necessary
* [ ] Updated affected documentation
* [x] Verified locally with `npm test`
* [x] Manually sanity tested running locally
* [x] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Squashed erroneous commits if necessary
* [ ] Re-tested if necessary

### CODE REVIEW

File: lib/twilio/audioplayer/audioplayer.ts [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+   * The source URL of the sound to play. When set, the currently playing sound will stop.
+   */
+  private _src: string = '';
+
+  /**
+   * The Factory to use to construct an XMLHttpRequest.
+   */
+  private _XMLHttpRequest: any;
+
+  get destination(): MediaStreamAudioDestinationNode { return this._destination; }
```
Comment (mhuynh5757) [inline lib/twilio/audioplayer/audioplayer.ts:99]: Out of scope, but I prefer not to use `getter` and `setter` methods. It obfuscates the possibility of side-effects when accessing members.

File: lib/twilio/audioplayer/audioplayer.ts [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+  const request: XMLHttpRequest = new RequestFactory();
+  request.open('GET', src, true);
+  request.responseType = 'arraybuffer';
+
+  const event: any = await new Promise(resolve => {
+    request.addEventListener('load', resolve);
+    request.send();
+  });
+
+  // Safari uses a callback here instead of a Promise.
```
Comment (mhuynh5757) [inline lib/twilio/audioplayer/audioplayer.ts:353]: This makes me sad.

File: lib/twilio/audioplayer/deferred.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  readonly promise: Promise<T>;
+
+  private _reject: (e?: any) => void;
+  get reject() { return this._reject; }
+
+  private _resolve: (t?: T) => void;
+  get resolve() { return this._resolve; }
+
+  constructor() {
+    this.promise = new Promise((resolve, reject) => {
```
Comment (mhuynh5757) [inline lib/twilio/audioplayer/deferred.ts:17]: Can you pass the type parameter to the Promise here so we don't have to typecast the `Deferred` objects in the main class?

**Review by mhuynh5757 — APPROVED** [review]

**Comment by charliesantos** [comment]
Got +1 from @mhuynh5757 https://twilio.slack.com/archives/C042FCK9WSD/p1686858451543789

**Pull request merged**